23 desuckify security

thatandromeda opened the pull request:
## Status
**READY**

#### What does this PR do?
We need to ensure that we don't accidentally enable fake auth in production. This checks on the Heroku app name as well as the `FAKE_AUTH_ENABLED` env variable.

#### Helpful background context (if appropriate)

#### How can a reviewer manually see the effects of these changes?

#### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TI-23

#### Screenshots (if appropriate)

#### Todo:
- [ ] Tests
- [ ] Documentation
- [ ] Stakeholder approval

#### Requires Database Migrations?
NO

#### Includes new or updated dependencies?
YES (for some reason dotenv wasn't already there)

thatandromeda commented:
@JPrevost This isn't done but I had some questions and this seemed like the easiest way to ask them.

* Does using `Rails.configuration.fake_auth_enabled` make sense? Using `Rails.configuration` isn't a pattern we use elsewhere so I wonder if it feels too weird.
* If not that, what? I thought about making it a method in ApplicationController or ApplicationHelper, but we need to check it from OmniauthCallbacksController (does not inherit from ApplicationController) and also `config/initializers/devise.rb`, which I think means I can't make it later than application config.
* That said I don't know if application config is even testable (how would i check to see that env had been properly translated into config without reinitializing config, and how can I reinitialize in the middle of running tests?) A helper method would be way easier to test but I would not be able to include it in the Devise initialization (would I?)

Anyway, I'm hoping you know some magic Railsy way around all this.

JPrevost commented on app.json:
```diff
@@ -0,0 +1,83 @@
+  "name": "thing",
+  "scripts": {},
+  "env": {
+    "CERTIFICATE": {
```
We shouldn't move the certs or IDP, URN, and SP stuff to pr apps. It won't work as they won't be registered in touchstone.

thatandromeda replied:
👍 Removing them.

JPrevost commented on app.json:
```diff
@@ -0,0 +1,83 @@
+    "CERTIFICATE": {
+      "required": true
+    },
+    "HEROKU_APP_NAME": {
```
Hmmm. I thought `required: true` copied the value from the staging app which would not be desirable. Maybe I don't get it.

thatandromeda replied:
HEROKU_APP_NAME follows a different pattern, which is weird: https://devcenter.heroku.com/articles/github-integration-review-apps

JPrevost replied:
Ah, weird, but okay :)

JPrevost commented on Gemfile:
```diff
@@ -25,6 +25,7 @@ group :development, :test do
   gem 'capybara'
   gem 'selenium-webdriver'
   gem 'sqlite3'
+  gem 'dotenv'
```
`heroku local` already loads stuff from `.env`. What are we gaining with `dotenv` gem? I'm not opposed to this if you find it better, but we will need to be careful as it can get weird with test envs if we aren't careful.

JPrevost commented on config/application.rb:
```diff
@@ -14,5 +14,14 @@ class Application < Rails::Application
     # Application configuration should go into files in config/initializers
     # -- all .rb files in that directory are automatically loaded.
+
+    Dotenv.load  # We need this to perform the checks below.
```
Oh. Hmmm. I'm not sure I love this. I'll get back to you but I feel like loading ENV outside of the normal way is asking for trouble longterm. The `dotenv` gem usually just loads without this anyway so it's possible this is not necessary?

JPrevost commented on config/application.rb:
```diff
@@ -14,5 +14,14 @@ class Application < Rails::Application
     # -- all .rb files in that directory are automatically loaded.
+
+    Dotenv.load  # We need this to perform the checks below.
+    review_app_pattern = /^library-thesis-dropbox-(staging-)?pr-[\d]+$/
```
All PR builds should be based on the staging app so the parens should be safe to remove.

JPrevost commented:
@thatandromeda I added a commit that moves the configuration to an initializer. I'm sorry I didn't just provide feedback to suggest doing that, but I wasn't entirely sure whether it would work out and honestly was not as straight forward as I initially assumed it would be. I moved to a module to allow for easier testing and am using a gem I've been meaning to try out to modify ENV in tests in a clean way. Thoughts are very welcome and again sorry for providing code rather than coherent thoughts on your PR :)

JPrevost commented:
@thatandromeda also, I removed dotenv as I couldn't get it to reliably work without calling `Dotenv.load` explicitly which I think would have errored fatally in production unless we included the gem outside of the development group...which I'm not sure I love. Maybe who care? I haven't decided. Either way, `heroku local` currently works as expected. I do want to discuss (soonish?) using `dotenv` vs `heroku local` as our standard for rails apps as not everything will target heroku but there is no good reason to not have everything using `.env` files in development.

thatandromeda commented on app/models/concerns/fake_auth_config.rb:
```diff
@@ -0,0 +1,26 @@
+
+  # Checks to make sure the application is not the staging or production
+  # instance via a heroku ENV.
+  # In development env we always return `true` to avoid having to set a fake app
```
👍

thatandromeda commented on config/initializers/0_fake_auth_configuration.rb:
```diff
@@ -0,0 +1,5 @@
+# This configuration needs to load before the devise initializer so this value
```
👍 Good thing to document in a comment.

thatandromeda commented:
@JPrevost I started the conversation in code rather than English; it's perfectly legit for you to continue it that way.

Your changes look 🌈 . I've rebased onto updated master and squashed - this is ready for review.
